core/camel-base-engine/src/main/java/org/apache/camel/impl/engine/DefaultRoutesBuilderLoader.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.impl.engine;
+
+import java.util.Optional;
+
+import org.apache.camel.CamelContext;
+import org.apache.camel.CamelContextAware;
+import org.apache.camel.ExtendedCamelContext;
+import org.apache.camel.RoutesBuilder;
+import org.apache.camel.spi.FactoryFinder;
+import org.apache.camel.spi.Resource;
+import org.apache.camel.spi.RoutesBuilderLoader;
+import org.apache.camel.util.ObjectHelper;
+import org.apache.camel.util.StringHelper;
+
+public class DefaultRoutesBuilderLoader implements RoutesBuilderLoader, CamelContextAware {
+    private CamelContext camelContext;
+
+    public DefaultRoutesBuilderLoader() {
+    }
+
+    public DefaultRoutesBuilderLoader(CamelContext camelContext) {
+        this.camelContext = camelContext;
+    }
+
+    @Override
+    public CamelContext getCamelContext() {
+        return camelContext;
+    }
+
+    @Override
+    public void setCamelContext(CamelContext camelContext) {
+        this.camelContext = camelContext;
+    }
+
+    @Override
+    public RoutesBuilder loadRoutesBuilder(Resource resource) throws Exception {
+        final String location = resource.getLocation();
+        final String language = getLanguage(location);
+
+        if (ObjectHelper.isEmpty(language)) {
+            throw new IllegalArgumentException("Unable to determine language for resource: " + location);
+        }
+
+        return getLoader(language).loadRoutesBuilder(resource);
+    }
+
+    private String getLanguage(String location) {
+        String answer = StringHelper.afterLast(location, ":", location);
+        answer = StringHelper.afterLast(answer, ".", answer);
+
+        return answer;
+    }
+
+    private RoutesBuilderLoader getLoader(String language) {
+        RoutesBuilderLoader answer = getCamelContext().getRegistry().lookupByNameAndType(language, RoutesBuilderLoader.class);
+
+        if (answer == null) {
+            final ExtendedCamelContext ecc = getCamelContext(ExtendedCamelContext.class);
+            final FactoryFinder finder = ecc.getFactoryFinder(RoutesBuilderLoader.LOADERS_FACTORY_PATH);
```

Comment:
Use bootstrap factory finder as we would load routes during bootstrapping/starting up camel